### PR TITLE
Add ich command

### PR DESCRIPTION
**Pull request by as0ler**
Add a new r2 command to export objc information in header format (as classdump does).

### CODE REVIEW

**Comment by Maijin** [comment]
Can you add test in cmd_info -> https://github.com/radare/radare2-regressions ?

File: libr/bin/format/objc/mach0_classes.c [hunk 1]
```diff
@@ -295,16 +299,16 @@ static void get_ivar_list_t(mach0_ut p, RBinFile *arch, RBinClass *klass) {
...
 				left = strlen (name) + 1;
 			} else {
-				name = malloc (left);
-				r_buf_read_at (arch->buf, r, (ut8 *)name, left);
+				type = malloc (left);
+				r_buf_read_at (arch->buf, r, (ut8 *)type, left);
 			}
-
-			R_FREE (name);
+      field->type = strdup(type);
```
Comment (XVilka) [inline libr/bin/format/objc/mach0_classes.c]: indent

File: libr/bin/format/objc/mach0_classes.c [hunk 2]
```diff
@@ -575,22 +579,24 @@ static void get_method_list_t(mach0_ut p, RBinFile *arch, char *class_name, RBin
...
-					free (name);
+				if (r_buf_read_at (arch->buf, r, (ut8 *)rtype, left) != left) {
+					free (rtype);
 					goto error;
 				}
-				name[left] = 0;
+				rtype[left] = 0;
 			}
-			R_FREE (name);
+      method->rtype = strdup(rtype);
```
Comment (XVilka) [inline libr/bin/format/objc/mach0_classes.c]: indent

File: libr/bin/format/objc/mach0_classes.c [hunk 3]
```diff
@@ -575,22 +579,24 @@ static void get_method_list_t(mach0_ut p, RBinFile *arch, char *class_name, RBin
...
 			}
 			if (bin->has_crypto) {
-				name = strdup ("some_encrypted_data");
-				left = strlen (name) + 1;
+				rtype = strdup ("some_encrypted_data");
+				left = strlen (rtype) + 1;
 			} else {
-				name = malloc (left + 1);
-				if (!name) {
+        left = 1;
```
Comment (XVilka) [inline libr/bin/format/objc/mach0_classes.c]: indent

File: libr/core/cbin.c [hunk 4]
```diff
@@ -2278,6 +2280,46 @@ static int bin_fields(RCore *r, int mode, int va) {
 	return true;
 }
 
+static char* get_rp (const char* rtype) {
+     char *rp = NULL;
+     if (r_str_cmp(rtype, "v", 1)) {
```
Comment (XVilka) [inline libr/core/cbin.c]: probably checking a char is faster than calling r_str_cmp() each time.

File: libr/core/cmd_info.c [hunk 5]
```diff
@@ -708,11 +709,14 @@ static int cmd_info(void *data, const char *input) {
 									r_cons_newline ();
 								}
 							}
+						} else if (input[1] == 'h' && obj) { // "ich"
+                mode = R_CORE_BIN_CLASSDUMP;
```
Comment (XVilka) [inline libr/core/cmd_info.c]: indent

**Comment by radare** [comment]
How to use this from rabin2?

File: libr/core/cbin.c (hunk 4 above)
Comment (radare) [inline libr/core/cbin.c]: just switch (*rtype) { like @XVilka says

File: libr/core/cbin.c [hunk 6]
```diff
@@ -2278,6 +2280,46 @@ static int bin_fields(RCore *r, int mode, int va) {
...
+        } else if (r_str_cmp(rtype, "f", 1)) {
+            rp = strdup("float");
+        } else if (r_str_cmp(rtype, "d", 1)) {
+            rp = strdup("double");
+        } else if (r_str_cmp(rtype, "D", 1)) {
+            rp = strdup("long double");
+        } else if (r_str_cmp(rtype, "B", 1)) {
+            rp = strdup("bool");
+        } else if (r_str_cmp(rtype, "#", 1)) {
+            rp = strdup("Class");
```
Comment (radare) [inline libr/core/cbin.c]: use tabs to indent and add missing space before (

File: libr/core/cbin.c [hunk 7]
```diff
@@ -2391,6 +2433,25 @@ static int bin_classes(RCore *r, int mode) {
 				}
 			}
 			r_cons_printf ("]}");
+		} else if (IS_MODE_CLASSDUMP (mode)) { 
+      char *rp = NULL;
```
Comment (radare) [inline libr/core/cbin.c]: use tabs

File: libr/core/cbin.c [hunk 8]
```diff
@@ -2391,6 +2433,25 @@ static int bin_classes(RCore *r, int mode) {
...
+        r_list_foreach (c->fields, iter2, f) {
+            if (f->name && r_regex_match("ivar","e", f->name)) {
+              r_cons_printf ("\t%s %s\n", f->type, f->name);
+            }
+        }
+        r_cons_printf("}\n");
+        r_list_foreach (c->methods, iter3, sym) {
+          if (sym->rtype && !r_str_cmp(sym->rtype,"@", 1)) {
+            rp = get_rp(sym->rtype);
+            r_cons_printf ("%s (%s) %s\n", r_str_cmp(sym->type,"METH",4) ? "-": "+", rp, sym->dname? sym->dname: sym->name);
```
Comment (radare) [inline libr/core/cbin.c]: use strcmp, sym->type shoulkdnt be null and compilers know how to optimize that, r_str_cmp cant be optimized

File: libr/include/r_core.h [hunk 9]
```diff
@@ -458,13 +458,14 @@ R_API void r_core_sysenv_help(const RCore* core);
...
 /* bin.c */
-#define R_CORE_BIN_PRINT	0x000
+#define R_CORE_BIN_PRINT	0x000 
 #define R_CORE_BIN_RADARE	0x001
 #define R_CORE_BIN_SET		0x002
 #define R_CORE_BIN_SIMPLE	0x004
-#define R_CORE_BIN_JSON         0x008
-#define R_CORE_BIN_ARRAY 	0x010
+#define R_CORE_BIN_JSON       0x008
+#define R_CORE_BIN_ARRAY 	    0x010
```
Comment (radare) [inline libr/include/r_core.h]: wtf is this indent mess

File: libr/include/r_bin.h [hunk 10]
```diff
@@ -473,6 +474,7 @@ typedef struct r_bin_field_t {
 	int size;
 	ut32 visibility;
 	char *name;
+  char *type;
```
Comment (radare) [inline libr/include/r_bin.h]: tabs

File: libr/core/cbin.c [hunk 11]
```diff
@@ -2281,43 +2281,61 @@ static int bin_fields(RCore *r, int mode, int va) {
...
-            rp = strdup("bool");
-        } else if (r_str_cmp(rtype, "#", 1)) {
-            rp = strdup("Class");
-        } else {
-            rp = strdup("unknown");
-        }
-  return rp;
+	char *rp = NULL;
+	switch(rtype[1]) {
+		case 'v':
```
Comment (radare) [inline libr/core/cbin.c]: Case goes at the same column as switch

File: libr/bin/format/objc/mach0_classes.c [hunk 12]
```diff
@@ -465,11 +469,13 @@ static void get_method_list_t(mach0_ut p, RBinFile *arch, char *class_name, RBin
 	mach0_ut r;
 	ut32 offset, left, i;
 	char *name = NULL;
+  char *rtype = NULL;
```
Comment (radare) [inline libr/bin/format/objc/mach0_classes.c]: Wrong indent

File: libr/bin/format/objc/mach0_classes.c [hunk 13]
```diff
@@ -465,11 +469,13 @@ static void get_method_list_t(mach0_ut p, RBinFile *arch, char *class_name, RBin
...
 	int len;
 	bool bigendian;
 	ut8 sml[sizeof (struct MACH0_(SMethodList))] = {0};
 	ut8 sm[sizeof (struct MACH0_(SMethod))] = {0};
 
+
```
Comment (radare) [inline libr/bin/format/objc/mach0_classes.c]: Two empty lines

File: libr/core/cbin.c [hunk 14]
```diff
@@ -2278,6 +2280,64 @@ static int bin_fields(RCore *r, int mode, int va) {
 	return true;
 }
 
+static char* get_rp (const char* rtype) {
+	char *rp = NULL;
+	switch(rtype[1]) {
+		case 'v':
```
Comment (radare) [inline libr/core/cbin.c]: Wrong indent

File: libr/core/cbin.c [hunk 15]
```diff
@@ -2278,6 +2280,64 @@ static int bin_fields(RCore *r, int mode, int va) {
...
+			rp = strdup("void");
+			break;
+		case 'c':
+			rp = strdup("char");
```
Comment (radare) [inline libr/core/cbin.c]: Missing space before all (

File: libr/core/cbin.c [hunk 16]
```diff
@@ -2391,11 +2451,30 @@ static int bin_classes(RCore *r, int mode) {
 				}
 			}
 			r_cons_printf ("]}");
+		} else if (IS_MODE_CLASSDUMP (mode)) { 
+			char *rp = NULL;
+			if (c) {
+				//TODO -> Print Superclass
+				r_cons_printf("@interface %s :  \n{\n", c->name);
```
Comment (radare) [inline libr/core/cbin.c]: Missing space before (

File: libr/core/cbin.c [hunk 17]
```diff
@@ -2391,11 +2451,30 @@ static int bin_classes(RCore *r, int mode) {
...
+					}
+				}
+				r_cons_printf("@end\n");
+			}
 		} else {
 			int m = 0;
 			r_cons_printf ("0x%08"PFMT64x" [0x%08"PFMT64x" - 0x%08"PFMT64x"] (sz %"PFMT64d") class %d %s",
 				c->addr, at_min, at_max, (at_max - at_min), c->index, c->name);
-			if (c->super) {
+		if (c->super) {
```
Comment (radare) [inline libr/core/cbin.c]: Wrong indent

File: libr/core/cbin.c [hunk 18]
```diff
@@ -2417,6 +2496,7 @@ static int bin_classes(RCore *r, int mode) {
 	return true;
 }
 
+
```
Comment (radare) [inline libr/core/cbin.c]: Two empty lines

File: libr/include/r_bin.h [hunk 19]
```diff
@@ -414,6 +414,7 @@ typedef struct r_bin_symbol_t {
 	const char *forwarder;
 	const char *bind;
 	const char *type;
+  const char *rtype;
```
Comment (radare) [inline libr/include/r_bin.h]: Wrong indent

File: libr/core/cbin.c [hunk 20]
```diff
@@ -2391,6 +2451,25 @@ static int bin_classes(RCore *r, int mode) {
...
 			}
 			r_cons_printf ("]}");
+		} else if (IS_MODE_CLASSDUMP (mode)) { 
+			char *rp = NULL;
+			if (c) {
+				//TODO -> Print Superclass
+				r_cons_printf ("@interface %s :  \n{\n", c->name); 
+				r_list_foreach (c->fields, iter2, f) {
+					if (f->name && r_regex_match ("ivar","e", f->name)) {
+						r_cons_printf ("\t%s %s\n", f->type, f->name);
```
Comment (radare) [inline libr/core/cbin.c]: do not use \t, only spaces

File: libr/core/cbin.c [hunk 21]
```diff
@@ -2391,6 +2451,25 @@ static int bin_classes(RCore *r, int mode) {
...
+					}
+				}
+				r_cons_printf ("}\n");
+				r_list_foreach (c->methods, iter3, sym) {
+					if (sym->rtype && (strncmp(sym->rtype,"@",1) != 0)) {
```
Comment (radare) [inline libr/core/cbin.c]: just sym->rtype[0] == '@'

Comment (radare) [inline libr/core/cbin.c]: cleaner, faster and more readable

File: libr/core/cbin.c [hunk 22]
```diff
@@ -2391,6 +2451,25 @@ static int bin_classes(RCore *r, int mode) {
...
+						rp = get_rp (sym->rtype);
+						r_cons_printf ("%s (%s) %s\n", strncmp (sym->type,"METH",4) == 0 ? "-": "+", rp, sym->dname? sym->dname: sym->name);
```
Comment (radare) [inline libr/core/cbin.c]: swap the condition and drop the == 0

File: libr/include/r_core.h [hunk 23]
```diff
@@ -458,13 +458,14 @@ R_API void r_core_sysenv_help(const RCore* core);
...
 /* bin.c */
-#define R_CORE_BIN_PRINT	0x000
+#define R_CORE_BIN_PRINT	0x000 
 #define R_CORE_BIN_RADARE	0x001
 #define R_CORE_BIN_SET		0x002
 #define R_CORE_BIN_SIMPLE	0x004
-#define R_CORE_BIN_JSON         0x008
-#define R_CORE_BIN_ARRAY 	0x010
-#define R_CORE_BIN_SIMPLEST 	0x020
+#define R_CORE_BIN_JSON		0x008
```
Comment (radare) [inline libr/include/r_core.h]: indent with spaces not tabs if u indent hte value

File: libr/core/cbin.c [hunk 24]
```diff
@@ -2278,6 +2280,64 @@ static int bin_fields(RCore *r, int mode, int va) {
...
+		rp = strdup ("double");
+		break;
+	case 'D': 
+		rp = strdup ("long double");
+		break;
+	case 'B':
+		rp = strdup ("bool");
+		break;
+	case '#':
+		rp = strdup ("Class");
```
Comment (radare) [inline libr/core/cbin.c]: uppercase class?

File: libr/bin/format/objc/mach0_classes.c [hunk 25]
```diff
@@ -575,22 +578,24 @@ static void get_method_list_t(mach0_ut p, RBinFile *arch, char *class_name, RBin
...
-					free (name);
+				if (r_buf_read_at (arch->buf, r, (ut8 *)rtype, left) != left) {
+					free (rtype);
 					goto error;
 				}
-				name[left] = 0;
+				rtype[left] = 0;
 			}
-			R_FREE (name);
+      			method->rtype = strdup(rtype);
```
Comment (radare) [inline libr/bin/format/objc/mach0_classes.c]: missing space before (

File: libr/bin/format/objc/mach0_classes.c [hunk 26]
```diff
@@ -295,16 +299,16 @@ static void get_ivar_list_t(mach0_ut p, RBinFile *arch, RBinClass *klass) {
...
 				left = strlen (name) + 1;
 			} else {
-				name = malloc (left);
-				r_buf_read_at (arch->buf, r, (ut8 *)name, left);
+				type = malloc (left);
+				r_buf_read_at (arch->buf, r, (ut8 *)type, left);
 			}
-
-			R_FREE (name);
+      			field->type = strdup(type);
```
Comment (radare) [inline libr/bin/format/objc/mach0_classes.c]: missing space before (

**Comment by radare** [comment]
ping

**Pull request merged**